File: basic.js
/*
 *
 *
 *   Binary Search
 */
// standard
const binarySearch = (list, target) => {
  let i = 0;
  let j = list.length - 1;
  let res = -1;
  while (i <= j) {
    let m = Math.floor(i + (j - i) / 2);
    if (list[m] < target) {
      i = m + 1;
    } else if (list[m] > target) {
      j = m - 1;
    } else {
      res = m;
      break;
    }
  }
  return res;
};

// the left most in targets
const binarySearchFirst = (list, target) => {
  let i = 0;
  let j = list.length - 1;
  let res = -1;
  while (i <= j) {
    let m = Math.floor(i + (j - i) / 2);
    if (list[m] < target) {
      i = m + 1;
    } else if (list[m] > target) {
      j = m - 1;
    } else {
      res = m;
      j = m - 1;
    }
  }
  return res;
};

// the right most in targets
const binarySearchLast = (list, target) => {
  let i = 0;
  let j = list.length - 1;
  let res = -1;
  while (i <= j) {
    let m = Math.floor(i + (j - i) / 2);
    if (list[m] < target) {
      i = m + 1;
    } else if (list[m] > target) {
      j = m - 1;
    } else {
      res = m;
      i = m + 1;
    }
  }
  return res;
};

// LowBound: the biggest that < target
const binarySearchLower = (list, target) => {
  let i = 0;
  let j = list.length - 1;
  let res = -1;
  while (i <= j) {
    let m = Math.floor(i + (j - i) / 2);
    if (list[m] < target) {
      i = m + 1;
      res = m;
    } else {
      j = m - 1;
    }
  }
  return res;
};

// HighBound: the smallest that > target
const binarySearchUpper = (list, target) => {
  let i = 0;
  let j = list.length - 1;
  let res = -1;
  while (i <= j) {
    let m = Math.floor(i + (j - i) / 2);
    if (list[m] <= target) {
      i = m + 1;
    } else {
      j = m - 1;
      res = m;
    }
  }
  return res;
};

/**
 *
 *
 *
 * Quick Sort
 */
// recursive
const quickSort = (list) => {
  if (list.length <= 1) return list;
  const { left, mid, right } = list.reduce(
    (a, c) => {
      if (c === list[0]) a.mid.push(c);
      else if (c < list[0]) a.left.push(c);
      else a.right.push(c);
      return a;
    },
    { left: [], mid: [], right: [] }
  );
  return [...quickSort(left), ...mid, ...quickSort(right)];
};

// in-place,  single boundary，[start, pivot] [p, ...end]
const quickSort2 = (list, start = 0, end = list.length - 1) => {
  if (start >= end) return;
  let p = start; // boundary, 从p开始都是value > pivot
  let pivot = list[end];
  for (let i = start; i <= end; ++i) {
    if (list[i] <= pivot) {
      swap(i, p);
      p++;
    }
  }
  // 注意 p-1 对应的pivot值(由最后一次循环的交换保证),
  quickSort2(list, start, p - 2); // 排除 pivot (如果p-1会导致 max call stack)
  quickSort2(list, p, end);
};

// quick select kth largest
const quickSelect = (left, right) => {
  let p = left,
    pivot = nums[right];
  for (let i = left; i < right; ++i) {
    // i<right 因为 最右边是 pivot（不动）
    if (nums[i] <= pivot) {
      swap(i, p);
      ++p;
    }
  }
  swap(p, right); // 最后把pivot放置到分界 p 就是 pivotIndex
  if (p < k) return quickSelect(p + 1, right);
  if (p > k) return quickSelect(left, p - 1);
  return nums[p];
};

// in-place,  double boundarys:  [start, i-1], [i==j/pivot], [j+1, end]
const quickSort3 = (list, start = 0, end = list.length - 1) => {
  if (start >= end) return;
  let i = start; // boundary, 从i往后value > pivot
  let j = end; // boundary, 从j往前value < pivot
  let pivot = list[end];
  while (i < j) {
    while (i < j && list[i] <= pivot) ++i;
    [list[i], list[j]] = [list[j], list[i]];
    while (i < j && list[j] >= pivot) --j;
    [list[i], list[j]] = [list[j], list[i]];
  }
  // 循环结束，i与j相等, 这个位置是属于pivot
  list[i] = pivot;

  quickSort3(list, start, i - 1);
  quickSort3(list, i + 1, end);
};

/**
 *
 *
 *
 * Union-Find
 */
class UnionFind {
  union(x, y) {
    const rootX = this.find(x);
    const rootY = this.find(y);
    if (rootX === rootY) return;
    this[rootX] = rootY;
  }
  find(x) {
    if (undefined === this[x]) this[x] = x;
    return this[x] === x ? x : this.find(this[x]);
  }
}

/**
 *
 *
 *
 * Trie
 */
class Trie {
  insert(arr) {
    let node = this;
    for (const e of arr) {
      node[e] = node[e] ?? new Trie();
      node = node[e];
    }
    node.isEnd = true;
  }

  has(arr) {
    let node = this;
    for (const e of arr) {
      if (!(e in node)) return false;
      node = node[e];
    }
    return node.isEnd;
  }

  startsWith(prefixArr) {
    let node = this;
    for (const e of prefixArr) {
      if (!(e in node)) return false;
      node = node[e];
    }
    return true;
  }
}

// const trie = new Trie();
// trie.insert(["a", "b", "c"]);
// console.log(trie);
// console.log(trie.has(["a", "b", "c"]), trie.has(["a", "b"]));
// console.log(trie.startsWith(["a", "b", "c"]), trie.startsWith(["a", "b", "a"]));

File: basic.rb
#   bundle install && bundle exec irb
#
#   require 'algorithms'
#   # [url](https://kanwei.github.io/algorithms)
#   include Containers
#
#   # -- test Containers::PriorityQueue  --
#   1. min-top heap
#         pq_min = PriorityQueue.new { _1 < _2 }
#         pq_min.push("Brad", 10)
#         pq_min.push("Yao", 10)
#         pq_min.next  # => peek "Brad"
#         pq_min.pop   # => remove and return "Brad"
#
#   2. max-top heap
#         pq_max = PriorityQueue.new { _1 > _2 }
#         pq_max.push("Brad", 10)
#         pq_max.push("Yao", 40)
#         pq_max.next  # => "Yao"

## binary search basic
def binary_search(nums, target)
  i = 0
  j = nums.length - 1
  while i <= j
    m = i + (j - i) / 2
    if nums[m] < target
      i = m + 1
    elsif nums[m] > target
      j = m - 1
    else
      return m
    end
  end
end
# puts found = binary_search([1,2,4,6,7,8,9], 8)

def binary_search_first(nums, target)
  i = 0
  j = nums.length - 1
  while i <= j
    m = i + (j - i) / 2
    if nums[m] < target
      i = m + 1
    else
      j = m - 1
      found = m if nums[m] == target
    end
  end
  found
end
# puts found = binary_search_first([1, 2, 3, 4, 4, 4, 4, 5, 5, 6, 6, 7], 4) # 3

def binary_search_last(nums, target)
  i = 0
  j = nums.length - 1
  while i <= j
    m = i + (j - i) / 2
    if nums[m] <= target
      i = m + 1
      found = m if nums[m] == target
    else
      j = m - 1
    end
  end
  found
end
# puts found = binary_search_last([1, 2, 3, 4, 4, 4, 4, 5, 5, 6, 6, 7], 4) # 6

def binary_search_lower(nums, target)
  i = 0
  j = nums.length - 1
  while i <= j
    m = i + (j - i) / 2
    if nums[m] < target
      found = m
      i = m + 1
    else
      j = m - 1
    end
  end
  found
end
# puts found = binary_search_lower([1, 2, 2, 2, 3, 4, 5, 5, 5, 6], 6) # 8

def binary_search_higher(nums, target)
  i = 0
  j = nums.length - 1
  while i <= j
    m = i + (j - i) / 2
    if nums[m] <= target
      i = m + 1
    else
      found = m
      j = m - 1
    end
  end
  found
end
# puts found = binary_search_higher([1, 2, 3, 4, 4, 4, 5, 5, 6], 4) # 6
#

# ==== quick sort  ====
def swap(list, i, j)
  list[i], list[j] = list[j], list[i]
end

# single pointer (easy to understand)
def quick_sort(list, s = 0, e = list.length - 1)
  return if s >= e

  p = s # boundary: from p all > pivot; before p all <= pivot
  pivot = list[e] # must be last to keep p-1 as pivot after last swap

  (s..e).each do |i|
    if list[i] <= pivot
      swap(list, i, p)
      p += 1
    end
  end

  quick_sort(list, s, p - 2) # exclude pivot (at index p-1), or might curse inifite recursion
  quick_sort(list, p, e)
end

def quick_sort_double_pointer(list, s = 0, e = list.length - 1)
  return if s >= e

  pivot = list[e]
  left = s
  right = e
  while left < right
    left += 1 while left < right && list[left] <= pivot # find an element > pivot
    right -= 1 while left < right && list[right] >= pivot # find an elment < pivot
    swap(list, left, right)
  end
  swap(list, left, e) # put the pivot into its final place
  quick_sort_double_pointer(list, s, left - 1)
  quick_sort_double_pointer(list, left + 1, e)
end

# ==== Trie tree ====
class Trie
  attr_accessor :is_end

  def initialize = @data = {}
  def [](key) = @data[key]

  def []=(key, value)
    @data[key] = value
  end

  def insert(chars)
    node = self
    for c in chars
      node[c] ||= Trie.new
      node = node[c]
    end
    node.is_end = true
  end

  def has(chars)
    node = self
    for c in chars
      return false unless node[c]

      node = node[c]
    end
    !!node.is_end
  end

  def prefix(chars)
    node = self
    for c in chars
      return false unless node[c]

      node = node[c]
    end
    true
  end
end

# trie = Trie.new
# trie.insert(["a", "b", "c"])
# puts trie.has(["a", "b", "c"]), trie.has(["a", "b"])
# puts trie.prefix(["a", "b", "c"]), trie.prefix(["a", "b", "a"])

class UnionFind
  attr_reader :root

  def initialize = @root = {}

  def union(x, y)
    root_x = find(x)
    root_y = find(y)
    @root[root_y] = root_x unless root_x == root_y
  end

  def find(x)
    return x if (@root[x] ||= x) == x # create root for x if not exists

    find(@root[x])
  end
end
